WASM runtime, BINARY chunk support

joshgoebel opened the pull request:
Closes #1259.  Closes #1693. Closes #1776. Closes #1777.  Closes #1778.  Closes #1780.

Building on the great work already done by @msx80.  What I've done here:

- rebased the existing work onto the latest TIC-80 codebase
- upgrading WASM3 to newer version
- flushed out the API almost fully (though there are some questions)
- implemented Bunnymark in Zig/WASM
- adds `import binary` and `export binary` commands
- wrote Zig language bindings for the TIC-80 WASM API
- `wat2wasm` stuff has been removed, could be added back as `script: wat` in the future

Unanswered questions (see #1259) / TODOS

- [x] What is the max size of a WASM binary? **256kb.** 
   #1780
- [x] How much memory should the runtime have access to? Currently 256kb.  128kb wasn't enough to even run Bunnies benchmark (1200 sprite limit) since now all 96kb is "reserved"...
- [x] Can core.memory.ram become a pointer?  **For now, may be changed later.**
   #1777
- [x] What exactly should the editor do (other than provide the magic `--script: wasm` comment...  **It provides some simple docs, license, README, etc..**
  #1775
- [ ] We may still need to further tweak some of the API functions specifically for WASM.
   #1778 
- [x] What to do about "projects"... I'm not sure they fit with the WASM workflow. **Binary chunks are currently not saved in projects.**
    #1775
- [x] BINARY chunk should be saved in multi-chunks if over 64kb, same as CODE
- [x] WASM demo cartridge 
  #1776 
- [x] move WASM bunny bench into `demo.c`, etc
- [ ] can we support `remap` callback on `map`? This one may be over my head.  WASM3 author thinks technically yes, but it sounds like a chore - and WASM3 doesn't provide the support pieces yet.
- [ ] Add BOOT API callback 
   #1773 
- [x] Figure out where to put per-languages TIC header libs
   #1771 

You can find the Zig project in `examples/zig_bunnymark`.  You'll need a very recent version of Zig to build it... someone is welcome to come in here and write a C/C++ version and library wrapper.  I'd recommend you use the same Makefile's that WASM4 is using and it should be simple.

joshgoebel commented:
Doesn't seem to build Windows... an issue with WABT?

```
D:\a\TIC-80\TIC-80\vendor\wabt\src\config.h(59,10): fatal error C1083: Cannot open include file: 'alloca.h': No such file or directory [D:\a\TIC-80\TIC-80\build\wabt.vcxproj]
1661
  stream.cc
1662
D:\a\TIC-80\TIC-80\vendor\wabt\src\config.h(59,10): fatal error C1083: Cannot open include file: 'alloca.h': No such file or directory [D:\a\TIC-80\TIC-80\build\wabt.vcxproj]
```

If we're agreed that `wat` doesn't need to be a part of this first pass I can rip this out for now... or if someone else could jump in and help out?

nesbox commented:
I agree, `wat` can be ripped out for now :)

joshgoebel commented:
```
In file included from /__w/TIC-80/TIC-80/vendor/wasm3/source/m3_exec_defs.h:11,
874
                 from /__w/TIC-80/TIC-80/src/api/wasm.c:36:
875
/__w/TIC-80/TIC-80/vendor/wasm3/source/m3_core.h:36:25: error: conflicting types for 'u32'
876
   36 | typedef uint32_t        u32;
877
      |                         ^~~
878
In file included from /__w/TIC-80/TIC-80/include/tic80.h:26,
879
                 from /__w/TIC-80/TIC-80/src/tic.h:25,
880
                 from /__w/TIC-80/TIC-80/src/api.h:25,
881
                 from /__w/TIC-80/TIC-80/src/core/core.h:25,
882
                 from /__w/TIC-80/TIC-80/src/api/wasm.c:23:
883
/__w/TIC-80/TIC-80/include/tic80_types.h:31:22: note: previous declaration of 'u32' was here
884
   31 | typedef unsigned int u32;
885
      |                      ^~~
```

Only the Raspberry baremetail build is failing now because on that platform evidently `uint32` and `unsigned int` are not the same... and WASM3 also aliases `u32`... Is there same way to be more explicit on Raspberry PI?  I'm not sure why we need these conditionals in `tic80_types.h`?

Help?

keithohara commented:
This looks awesome!

How do you run the zig example?

This didn't work:

```
 ../../build/bin/tic80 bench.tic
[string ";; title:  game title..."]:1: function arguments expected near 'title'
```

joshgoebel commented:
Are you running the branch with this specific change?  This is not released anywhere, and it's not merged.

You'll need `--` comments, let me update.

keithohara commented:
> Are you running the branch with this specific change? This is not released anywhere, and it's not merged.
> 
> You'll need `--` comments, let me update.

Pretty sure I am:

```
 git branch -a
* wasm_support
```

joshgoebel commented:
That should fix it... I rebased onto master and @nesbox changed the memory map out from underneath me... grrrrr. :-p  So now I've added 1024 bytes padding at the end until we sort that out.

@nesbox Once we start actually using the free RAM for compiled software we can't just add it randomly - that would break all compiled cartridges.  If that memory at the end is truly reserved instead of "free" then we need to document that.  

Do you want to just reserve the entire 12kb for your own use or should we just flag off another 1-2KB that's reserved and "do not touch"?

keithohara commented:
> That should fix it... I rebased onto master and @nesbox changed the memory map out from underneath me... grrrrr. :-p So now I've added 1024 bytes padding at the end until we sort that out.

Thanks, that did the trick!

joshgoebel commented:
Let me know if you have any questions!  I'm on Discord also.

joshgoebel commented:
Done:

https://github.com/nesbox/TIC-80/wiki/RAM

But now that means that no cartridges (other than WASM) have access to any actual device memory at all... that's a change... should we bump the core memory from 96kb to 128kb -giving users 32kb back for hardware hacking in non-WASM runtimes?  I've used that memory before for temporarily copying portions of the screen buffer and such things.  There are nice tricks you can do with this and bank switching... but if you take away the free RAM you leave no one anywhere to `memcpy`.  

Thoughts?

nesbox commented:
This memory was never free, it was always assumed that sooner or later it will be occupied as it develops, let's leave it as is.

joshgoebel commented:
Once this is ready is merging it as experimental (behind a build flag that's disabled by default) reasonable? I'd love to allow a few people to build and test all (after merging) the API bits without also (yet) being locked into binary compatible for life... ie if we need to change a few things early on I'd like to do it without fear of breaking published cartridges in the wild.

TLDR: _Can we merge it but also label it "beta" and still subject to change slightly?_

joshgoebel commented on src/core/core.c:
```diff
@@ -444,6 +447,13 @@ void tic_core_tick(tic_mem* tic, tic_tick_data* data)
+            // TODO: does where to fetch code from need to be a config option so this isn't hard
+            // coded for just a single langage? perhaps change it later when we have a second script
+            // engine that uses BINARY?
+            if (strcmp(config->name,"wasm")==0) {
```
Or we could move this into the `config` struct, `usesBinaryVsCode` or some such, but perhaps that's premature...

nesbox commented:
I tried to build it and got this screen
![image](https://user-images.githubusercontent.com/1101448/154911385-fb9da450-f9d4-4f7a-a7db-a9dc0342d690.png)

Something is broken, possibly the problem is somewhere in the cart loading.

joshgoebel commented:
What exactly did you do to get that screen?

nesbox commented:
Nothing special, just compiled the `wasm_support` branch and ran.

joshgoebel commented:
Hmm.... I pushed my latest rebase so we're using the exact same commits, but mine has been building and running without issue all along...

joshgoebel commented:
Can you pul the latest and try again the commit number showing in your build is quite a while back...

nesbox commented:
I see that CMake is changing `config.tic.dat` at compile time and I think that's the reason I got this black screen.
![image](https://user-images.githubusercontent.com/1101448/154958793-c4afb778-9461-443d-be31-c683bf94f711.png)

nesbox commented:
I did some refactoring with the Studio here 3dc2feacbeb2f1db8e7184eb548d381d7e690efc which touched a lot of project files, pls don't worry, your PR can be easily merged with it 😟

joshgoebel commented:
> I see that CMake is changing config.tic.dat at compile time and I think that's the reason I got this black screen.

I'm assuming that's your new refactoring things?  I haven't made any changes to CMakeList that should touch config.tic.dat that I'm aware of...?   If you need me to make any changes or do any specific work here, please let me know.

nesbox commented:
Finally, I merged the PR with the `main` branch here [wasm-merge](https://github.com/nesbox/TIC-80/tree/wasm-merge).
I decided to split the cart builders to `prj2cart` and `wasmp2cart`, also I fixed an error, instead `fread(buffer, size, 1, file);` should be `fread(buffer, *size, 1, file);`, after that everything worked and finally I saw a normally rendered TIC-80 screen.
Seems all work except barematal build, don't know how to fix it properly.
![image](https://user-images.githubusercontent.com/1101448/155152241-eaeec6f6-85e7-4ca0-88a4-31381047fa61.png)
https://github.com/nesbox/TIC-80/runs/5289437702?check_suite_focus=true#step:6:876

joshgoebel commented:
> I decided to split the cart builders to prj2cart and wasmp2cart

Yeah, if it was up to me tiny helpers like this would be in a more accessible language, not C, so I'm glad you decided to rejigger that stuff...

Does Raspberry pi really not define `uint32_t`, that seems strange...?

nesbox commented:
I fixed RPI build with the patch, don't see another better way.
So, seems we are ready to merge the PR, there are two ways to do it, I can merge [wasm-merge](https://github.com/nesbox/TIC-80/tree/wasm-merge) branch directly to the `main` and close the PR or we could merge it with your `wasm_support` fork and then properly push it to the main repo.
What do you think?

nesbox commented:
Ok, I merged it directly to the `main` aeabb55bef63a11014107d2d0a369ac7ec324530.
Pls do something with the PR.
